File: swift-wasm-webgl/Sources/BufferUsage.swift
import JavaScriptKit

enum BufferUsage {
    case StaticDraw
    case DynamicDraw
    case StreamDraw
    case StaticRead
    case DynamicRead
    case StreamRead
    case StaticCopy
    case DynamicCopy
    case StreamCopy

    func glValue(gl: JSValue) -> Int {
        switch self {
        case .StaticDraw:
            return Int(gl.STATIC_DRAW.number!)
        case .DynamicDraw:
            return Int(gl.DYNAMIC_DRAW.number!)
        case .StreamDraw:
            return Int(gl.STREAM_DRAW.number!)
        case .StaticRead:
            return Int(gl.STATIC_READ.number!)
        case .DynamicRead:
            return Int(gl.DYNAMIC_READ.number!)
        case .StreamRead:
            return Int(gl.STREAM_READ.number!)
        case .StaticCopy:
            return Int(gl.STATIC_COPPY.number!)
        case .DynamicCopy:
            return Int(gl.DYNAMIC_COPY.number!)
        case .StreamCopy:
            return Int(gl.STREAM_COPY.number!)
        }
    }
}

File: rust-webgl/src/graphics/buffer_usage.rs
use web_sys::WebGl2RenderingContext;

pub enum BufferUsage {
    StaticDraw,
    DynamicDraw,
    StreamDraw,
    StaticRead,
    DynamicRead,
    StreamRead,
    StaticCopy,
    DynamicCopy,
    StreamCopy,
}

impl BufferUsage {
    pub fn gl_usage(&self) -> u32 {
        match self {
            BufferUsage::StaticDraw => WebGl2RenderingContext::STATIC_DRAW,
            BufferUsage::DynamicDraw => WebGl2RenderingContext::DYNAMIC_DRAW,
            BufferUsage::StreamDraw => WebGl2RenderingContext::STREAM_DRAW,
            BufferUsage::StaticRead => WebGl2RenderingContext::STATIC_READ,
            BufferUsage::DynamicRead => WebGl2RenderingContext::DYNAMIC_READ,
            BufferUsage::StreamRead => WebGl2RenderingContext::STREAM_READ,
            BufferUsage::StaticCopy => WebGl2RenderingContext::STATIC_COPY,
            BufferUsage::DynamicCopy => WebGl2RenderingContext::DYNAMIC_COPY,
            BufferUsage::StreamCopy => WebGl2RenderingContext::STREAM_COPY,
        }
    }
}
